File: solutions/leetcode/contests/w-438/2-maximum_sum_with_at_most_k_elements.java
class Solution {
    public long maxSum(int[][] grid, int[] limits, int k) {
        List<Integer> values = new ArrayList<>();
        int n = grid.length;
        
        for (int i = 0; i < n; i++) {
            Arrays.sort(grid[i]); 
            for (int j = grid[i].length - 1, count = 0; j >= 0 && count < limits[i]; j--, count++) {
                values.add(grid[i][j]);
            }
        }
        
        Collections.sort(values, Collections.reverseOrder()); 
        
        long sum = 0;
        for (int i = 0; i < Math.min(k, values.size()); i++) {
            sum += values.get(i);
        }
        
        return sum;
    }
}

File: solutions/leetcode/contests/w-438/2-maximum_sum_with_at_most_k_elements.scala
object Solution {
    def maxSum(grid: Array[Array[Int]], limits: Array[Int], k: Int): Long = {
        val values = grid.zip(limits).flatMap { case (row, limit) =>
            row.sorted(Ordering[Int].reverse).take(limit)
        }
        values.sorted(Ordering[Int].reverse).take(k).map(_.toLong).sum
    }
}
